openquake/hazardlib/gsim/stafford_2022.py

```diff
@@ -0,0 +1,487 @@
+    """
+    Returns the basin depth scaling
+    """
+    return C["phi5"] * (1.0 - np.exp(-delta_z1pt0 /
+                                     CONSTANTS["phi6"]))
+def get_linear_site_term(C, ctx):
+    """
+    Returns the linear site scaling term
+    """
+    return C["phi1"] * np.log(ctx.vs30 / 1130).clip(-np.inf, 0.0)
+
+def get_nonlinear_site_term(C, ctx, y_ref):
+    """
+    Returns the nonlinear site term and the Vs-scaling factor (to be
+    used in the standard deviation model
+    """
+    vs = ctx.vs30.clip(-np.inf, 1130.0)
+    f_nl_scaling = C["phi2"] * (np.exp(C["phi3"] * (vs - 360.)) -
+                                np.exp(C["phi3"] * (1130. - 360.)))
+    f_nl = np.log((y_ref + C["phi4"]) / C["phi4"]) * f_nl_scaling
+    return f_nl, f_nl_scaling
+
+def get_ln_y_ref(T, C, ctx, mu_branch, adjust_c1, adjust_chm, adjust_c7, adjust_cg1):
+    """
+    Returns the ground motion on the reference rock.
+    """
+    delta_ztor = _get_centered_ztor(ctx)
+    return (get_stress_scaling(T, C, ctx, mu_branch, adjust_c1, adjust_chm, adjust_c7, adjust_cg1) +
+            get_magnitude_scaling(C, ctx.mag) +
+            get_source_scaling_terms(T, C, ctx, delta_ztor, mu_branch, adjust_c1, adjust_chm, adjust_c7, adjust_cg1) +
+            get_hanging_wall_term(C, ctx) +
+            get_geometric_spreading(C, ctx.mag, ctx.rrup) +
+            get_far_field_distance_scaling(T, C, ctx, mu_branch,adjust_c1, adjust_chm, adjust_c7, adjust_cg1) +
+            get_directivity(C, ctx))
+
+def _nl_sigma (C, ctx, ln_y_ref):
+    vs = ctx.vs30.clip(-np.inf, 1130.0)
+    NL = C['phi2'] * (np.exp(C['phi3'] * (vs - 360.0)) - np.exp(C['phi3'] * (1130.0 - 360.0))) * (np.exp(ln_y_ref) / (np.exp(ln_y_ref) + C['phi4']))
+    return NL
+
+def get_stddevs (T, ln_y_ref, C, ctx, sigma_branch):
+    """
+    Returns the standard deviation model described in equation 8.16 (Peter Stafford NSHM report)
+    """
+    tau = _between_event(T, ctx.mag)
+    phi_S2S = _between_station(T)
+    phi_SS = 0.39
+    NL = _nl_sigma(C, ctx, ln_y_ref)
+    phi = np.sqrt(phi_S2S**2 + (1.0 + NL)**2 * 0.736 * phi_SS**2 + 0.264 * phi_SS**2)
+    # nominal total standard deviation for the branch
+    sigma = np.sqrt((1.0 + NL)**2 * tau**2 + phi**2)
+
+    #perform the potential inflation of σ for epistemic uncertainty
+    neff = _neff_model(T)
+    nfac = (_empirical_sigma(T, ctx.rrup, ctx.mag) / _empirical_sigma(T, 80.0, 5.0))**2
+
+    ndof = neff / nfac
+    if sigma_branch == "Lower":
+        sigma = np.sqrt((sigma**2 / (ndof - 1)) * stats.chi2.ppf(0.1, ndof-1))
+    elif sigma_branch == "Central":
+        sigma = np.sqrt((sigma**2 / (ndof - 1)) * stats.chi2.ppf(0.5, ndof-1))
+    elif sigma_branch == "Upper":
+        sigma = np.sqrt((sigma**2 / (ndof - 1)) * stats.chi2.ppf(0.9, ndof-1))
+    else:
+        print("sigma branch not recognised: must be one of :Lower, :Central, :Upper")
+        sigma = np.NaN
+    return sigma, tau, phi
```

Comment:
@chrisdicaprio Working with this GMM I found the small issue here, which is that in this case that the total standard deviation contains a dependency on the nonlinear amplification term then `tau` should be returned as `(1.0 + NL) ** 2 * tau ** 2` (see the original Chiou & Youngs implementation here: https://github.com/gem/oq-engine/blob/master/openquake/hazardlib/gsim/chiou_youngs_2014.py#L370). This doesn't affect the classical PSHA calculations as they rely only on `sigma`, but it does affect the event-based risk calculations where between- and within-event residuals are sampled separately, so the total variability of the GMM would be different in the event-based case compared to the classical PSHA case.